index.js

```diff
@@ -219,88 +204,92 @@ async function handlePaidScheduledRewards (req, res, redis, signerAddresses, red
   }
 
   logger.info(`Marking scheduled rewards of ${body.participants.length} participants as paid`)
-  let updatedRewards
+  const scheduledRewardsDeltas = body.rewards.map(r => BigInt(r) * -1n)
 
-  const lock = await redlock.lock('lock:rewards', 60_000)
+  const pgClient = await pgPool.connect()
   try {
-    const currentRewards = (await redis.hmget('rewards', ...body.participants)).map(amount => {
-      return BigInt(amount || '0')
-    })
-    updatedRewards = body.rewards.map((amount, i) => {
-      const updated = currentRewards[i] - BigInt(amount)
-      httpAssert(
-        updated >= 0n,
-        400,
-        `Scheduled rewards of ${body.participants[i]} would become negative (current=${currentRewards[i]} paid=${amount})`
-      )
-      return updated
-    })
-
-    const tx = redis.multi()
-    tx.hset(
-      'rewards',
-      Object.fromEntries(body.participants.map((address, i) => ([
-        address,
-        String(updatedRewards[i])
-      ])))
-    )
-    for (let i = 0; i < body.participants.length; i++) {
-      addLogJSON(tx, {
-        timestamp,
-        address: body.participants[i],
-        scheduledRewardsDelta: String(BigInt(body.rewards[i]) * -1n)
-      })
+    await pgClient.query('BEGIN')
+    await pgClient.query(`
+      UPDATE scheduled_rewards
+      SET
+        amount = scheduled_rewards.amount + bulk.amount
+      FROM (
+        SELECT *
+          FROM
+            UNNEST($1::TEXT[], $2::NUMERIC[])
+          AS t(address, amount)
+      ) AS bulk
+      WHERE scheduled_rewards.address = bulk.address
+    `, [body.participants, scheduledRewardsDeltas])
+    await pgClient.query(`
+      INSERT INTO logs (address, scheduled_rewards_delta)
+      VALUES (UNNEST($1::TEXT[]), UNNEST($2::NUMERIC[]))
+    `, [body.participants, scheduledRewardsDeltas])
+    await pgClient.query(`
+      DELETE FROM logs WHERE timestamp < NOW() - INTERVAL '30 days'
+    `)
+    await pgClient.query('COMMIT')
+  } catch (err) {
+    await pgClient.query('ROLLBACK')
+    if (err.constraint === 'amount_not_negative') {
+      httpAssert.fail(400, `Scheduled rewards would become negative: ${err.detail}`)
     }
-    await tx.exec()
+    throw err
   } finally {
-    await lock.unlock()
+    pgClient.release()
   }
 
+  logger.info(`Increased scheduled rewards of ${body.participants.length} participants`)
+
   logger.info(`Merked scheduled rewards of ${body.participants.length} participants as paid`)
+
+  const { rows } = await pgPool.query(`
+    SELECT address, amount
+    FROM scheduled_rewards
+    WHERE address = ANY($1)  
+  `, [body.participants])
```

Comment:
Ditto - I think this second query can be avoided by modifying the first query `UPDATE scheduled_rewards` to return the new balance of each affected participant.

Reply:
also here inside a transaction